pkg/incentives/contract/redistribution.go

```diff
@@ -0,0 +1,254 @@
+	IsWinner(context.Context) (bool, error)
+	Claim(context.Context) error
+	Commit(context.Context, []byte) error
+	Reveal(context.Context, uint8, []byte, []byte) error
+}
+
+type Service struct {
+	overlay                   common.Hash
+	logger                    log.Logger
+	txService                 transaction.Service
+	incentivesContractAddress common.Address
+}
+
+func New(
+	overlay common.Hash,
+	logger log.Logger,
+	txService transaction.Service,
+	incentivesContractAddress common.Address,
+) *Service {
+
+	s := &Service{
+		overlay:                   overlay,
+		logger:                    logger.WithName(loggerName).Register(),
+		txService:                 txService,
+		incentivesContractAddress: incentivesContractAddress,
+	}
+	return s
+}
+
+func (s *Service) IsPlaying(ctx context.Context, depth uint8) (bool, error) {
+	callData, err := incentivesContractABI.Pack("isParticipatingInUpcomingRound", s.overlay, depth)
+	if err != nil {
+		return false, err
+	}
+
+	result, err := s.txService.Call(ctx, &transaction.TxRequest{
+		To:   &s.incentivesContractAddress,
+		Data: callData,
+	})
+	if err != nil {
+		return false, err
+	}
+
+	results, err := incentivesContractABI.Unpack("isParticipatingInUpcomingRound", result)
+	if err != nil {
+		return false, err
+	}
+
+	return results[0].(bool), nil
+}
+
+func (s *Service) IsWinner(ctx context.Context) (isWinner bool, err error) {
+	//winnerSeed, truthSeed, err := s.checkIsWinning(ctx)
+	//if err != nil {
+	//	return false, false, err
+	//}
+
+	callData, err := incentivesContractABI.Pack("isWinner", s.overlay)
+	if err != nil {
+		return false, err
+	}
+
+	result, err := s.txService.Call(ctx, &transaction.TxRequest{
+		To:   &s.incentivesContractAddress,
+		Data: callData,
+	})
+	if err != nil {
+		return false, err
+	}
+
+	results, err := incentivesContractABI.Unpack("isWinner", result)
+	if err != nil {
+		return false, err
+	}
+
+	return results[0].(bool), nil
+}
+
+func (s *Service) Claim(ctx context.Context) error {
+	callData, err := incentivesContractABI.Pack("claim")
+
+	request := &transaction.TxRequest{
+		To:   &s.incentivesContractAddress,
+		Data: callData,
+	}
+
+	txHash, err := s.txService.Send(ctx, request)
+	if err != nil {
+		return err
+	}
+
+	receipt, err := s.txService.WaitForReceipt(ctx, txHash)
+	if err != nil {
+		return err
+	}
+
+	if receipt.Status == 0 {
+		return transaction.ErrTransactionReverted
+	}
+
+	return nil
+}
+
+func (s *Service) Commit(ctx context.Context, obfusHash []byte) error {
+	callData, err := incentivesContractABI.Pack("commit", s.overlay, obfusHash)
+
+	request := &transaction.TxRequest{
+		To:   &s.incentivesContractAddress,
+		Data: callData,
+	}
+
+	txHash, err := s.txService.Send(ctx, request)
+	if err != nil {
+		return err
+	}
+
+	receipt, err := s.txService.WaitForReceipt(ctx, txHash)
+	if err != nil {
+		return err
+	}
+
+	if receipt.Status == 0 {
+		return transaction.ErrTransactionReverted
+	}
+
+	return nil
```

Comment:
consider reusing code

following block is repeated in many methods:

```
request := &transaction.TxRequest{
		To:   &s.incentivesContractAddress,
		Data: callData,
	}

	txHash, err := s.txService.Send(ctx, request)
	if err != nil {
		return err
	}

	receipt, err := s.txService.WaitForReceipt(ctx, txHash)
	if err != nil {
		return err
	}

	if receipt.Status == 0 {
		return transaction.ErrTransactionReverted
	}
```

 which could be moved to separate function 
 
 ```
 s.SendAndWait(callData)
 ```